### PR TITLE
ENG-4560 +ENG-4561 +ENG-4562 +ENG-4563 +ENG-4564 +ENG-4565 +ENG-4570 Multitenancy implementation

### PR DESCRIPTION
**Pull request by entgigi**

### CODE REVIEW

File: engine/src/test/java/com/agiletec/aps/system/EntThreadLocalTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+class EntThreadLocalTest {
+
+    @Test
+    void reqCtxThreadShouldInitInsertGet() {
+        final String keytest = "keytest";
+        final String valuetest = "valuetets";
```
Comment (zonia3000) [inline engine/src/test/java/com/agiletec/aps/system/EntThreadLocalTest.java]: typo

File: engine/src/test/java/com/agiletec/aps/system/EntThreadLocalTest.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        EntThreadLocal.clear();
+        EntThreadLocal.set(keytest, valuetest);
+        EntThreadLocal.clear();
+        Assertions.assertNull(EntThreadLocal.get(keytest));
+    }
+
+    @Test
+    void reqCtxThreadShouldRemoveOrDestroy() {
+        final String keytest = "keytest";
+        final String valuetest = "valuetets";
```
Comment (zonia3000) [inline engine/src/test/java/com/agiletec/aps/system/EntThreadLocalTest.java]: typo

File: engine/src/test/java/com/agiletec/aps/util/ApsTenantApplicationUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        Optional<String> tanant = ApsTenantApplicationUtils.getTenant();
+        Assertions.assertThat(tanant).isNotPresent();
+
+        ApsTenantApplicationUtils.setTenant("tCode");
+        tanant = ApsTenantApplicationUtils.getTenant();
+        Assertions.assertThat(tanant.get()).isNotEmpty().isEqualTo("tCode");
+
+        ApsTenantApplicationUtils.removeTenant();
+        tanant = ApsTenantApplicationUtils.getTenant();
+        Assertions.assertThat(tanant).isNotPresent();
```
Comment (zonia3000) [inline engine/src/test/java/com/agiletec/aps/util/ApsTenantApplicationUtilsTest.java]: typo

File: engine/src/test/java/org/entando/entando/aps/system/services/tenants/TenantConfigTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.stream.Collectors;
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+class TenantConfigTest {
+
+    @Test
+    void shouldConstructorFromMapAndCopyConstructorWorkFine(){
+
+        Map<String,String> map = Map.of("tenantCode","CodeTanant1","kCEnabled", "true", "kcClientId", "tenat1ClientId");
```
Comment (zonia3000) [inline engine/src/test/java/org/entando/entando/aps/system/services/tenants/TenantConfigTest.java]: typos: tenat1, Tanant

File: engine/src/test/java/org/entando/entando/aps/system/services/tenants/TenantConfigTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        Assertions.assertThat(dbMaxIdle).isEqualTo(2);
+        Assertions.assertThat(dbMaxWaitMillis).isEqualTo(100000);
+        Assertions.assertThat(dbInitialSize).isEqualTo(12);
+    }
+
+    @Test
+    void shouldDefaultGetWorkFine(){
+        Map<String,String> map = Map.of("tenantCode","1",
+                "kcEnabled","True",
+        "kcAuthUrl","3",
```
Comment (zonia3000) [inline engine/src/test/java/org/entando/entando/aps/system/services/tenants/TenantConfigTest.java]: indentation

File: keycloak-plugin/src/test/java/org/entando/entando/keycloak/services/KeycloakConfigurationTest.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                "kcDefaultAuthorizations", "");
+
+        ApsTenantApplicationUtils.setTenant("my-test-tenant");
+        KeycloakConfiguration kc = new KeycloakConfiguration();
+        TenantConfig tc = new TenantConfig(map);
+        Mockito.when(tenantManager.getConfig("my-test-tenant")).thenReturn(tc);
+        kc.setTenantManager(tenantManager);
+
+        Assertions.assertThat(kc.isEnabled()).isFalse();
+        org.junit.jupiter.api.Assertions.assertThrows(IllegalArgumentException.class, () -> kc.getRealm());
```
Comment (zonia3000) [inline keycloak-plugin/src/test/java/org/entando/entando/keycloak/services/KeycloakConfigurationTest.java]: Better to use only one assertion library

File: keycloak-plugin/src/test/java/org/entando/entando/keycloak/services/KeycloakServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+
+
+    @Test
+    void shouldThrowException() throws Exception {
+        Mockito.when(keycloakConfiguration.getAuthUrl()).thenReturn("http://localhost/auth");
+        Mockito.when(keycloakConfiguration.getRealm()).thenReturn("my-realm");
+
+        Mockito.when(openIDConnectService.authenticateAPI()).thenThrow(new OidcException(new Exception("test-message")));
+        org.junit.jupiter.api.Assertions.assertThrows(RuntimeException.class, () ->keycloakService.listUsers());
```
Comment (zonia3000) [inline keycloak-plugin/src/test/java/org/entando/entando/keycloak/services/KeycloakServiceTest.java]: Also here

**Review by eugeniosant — APPROVED** [review]

**Review by zonia3000 — APPROVED** [review]

**Pull request merged**